File: src/skill.c
#include "skill.h"
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <dirent.h>
#include <sys/stat.h>

// Directories to search for skills
static const char *SKILL_SEARCH_PATHS[] = {
    "skills",
    ".agents/skills",
    ".claude/skills",
    ".cursor/skills",
    ".cline/skills",
    ".codex/skills",
    NULL
};

static Skill *skill_alloc(void) {
    Skill *skill = spm_malloc(sizeof(Skill));
    skill->name = NULL;
    skill->description = NULL;
    skill->path = NULL;
    skill->skill_file = NULL;
    return skill;
}

void skill_free(Skill *skill) {
    if (!skill) return;
    spm_free(skill->name);
    spm_free(skill->description);
    spm_free(skill->path);
    spm_free(skill->skill_file);
    spm_free(skill);
}

void skill_list_free(SkillList *list) {
    if (!list) return;
    for (int i = 0; i < list->count; i++) {
        spm_free(list->skills[i].name);
        spm_free(list->skills[i].description);
        spm_free(list->skills[i].path);
        spm_free(list->skills[i].skill_file);
    }
    spm_free(list->skills);
    spm_free(list);
}

static void skill_list_add(SkillList *list, Skill *skill) {
    if (list->count >= list->capacity) {
        list->capacity = list->capacity == 0 ? 8 : list->capacity * 2;
        list->skills = spm_realloc(list->skills, list->capacity * sizeof(Skill));
    }
    list->skills[list->count++] = *skill;
    spm_free(skill);
}

// Parse YAML frontmatter from SKILL.md
// Format:
// ---
// name: skill-name
// description: Some description
// ---
Skill *parse_skill_file(const char *skill_md_path) {
    FILE *fp = fopen(skill_md_path, "r");
    if (!fp) {
        log_debug("Cannot open: %s", skill_md_path);
        return NULL;
    }

    Skill *skill = skill_alloc();
    char line[4096];
    bool in_frontmatter = false;
    bool found_frontmatter = false;

    while (fgets(line, sizeof(line), fp)) {
        char *trimmed = str_trim(line);

        if (strcmp(trimmed, "---") == 0) {
            if (!in_frontmatter) {
                in_frontmatter = true;
                continue;
            } else {
                // End of frontmatter
                found_frontmatter = true;
                break;
            }
        }

        if (in_frontmatter) {
            // Parse key: value
            char *colon = strchr(trimmed, ':');
            if (colon) {
                *colon = '\0';
                char *key = str_trim(trimmed);
                char *value = str_trim(colon + 1);

                // Remove quotes if present
                size_t vlen = strlen(value);
                if (vlen >= 2 && ((value[0] == '"' && value[vlen-1] == '"') ||
                                  (value[0] == '\'' && value[vlen-1] == '\''))) {
                    value[vlen-1] = '\0';
                    value++;
                }

                if (strcmp(key, "name") == 0) {
                    skill->name = str_dup(value);
                } else if (strcmp(key, "description") == 0) {
                    skill->description = str_dup(value);
                }
            }
        }
    }

    fclose(fp);

    if (!found_frontmatter || !skill->name) {
        // If no name in frontmatter, derive from directory name
        if (!skill->name) {
            // Get parent directory name
            char *path_copy = str_dup(skill_md_path);
            char *last_slash = strrchr(path_copy, '/');
            if (last_slash) {
                *last_slash = '\0';
                char *dir_slash = strrchr(path_copy, '/');
                if (dir_slash) {
                    skill->name = str_dup(dir_slash + 1);
                }
            }
            spm_free(path_copy);
        }
    }

    if (!skill->name) {
        skill_free(skill);
        return NULL;
    }

    skill->skill_file = str_dup(skill_md_path);

    // Set path to the directory containing SKILL.md
    char *path_copy = str_dup(skill_md_path);
    char *last_slash = strrchr(path_copy, '/');
    if (last_slash) {
        *last_slash = '\0';
        skill->path = path_copy;
    } else {
        skill->path = str_dup(".");
        spm_free(path_copy);
    }

    return skill;
}

// Read entire file into a freshly-allocated, NUL-terminated buffer.
// Returns NULL on any error.
static char *read_file_to_string(const char *path) {
    FILE *fp = fopen(path, "rb");
    if (!fp) return NULL;

    if (fseek(fp, 0, SEEK_END) != 0) {
        fclose(fp);
        return NULL;
    }
    long len = ftell(fp);
    if (len < 0) {
        fclose(fp);
        return NULL;
    }
    rewind(fp);

    char *buf = spm_malloc((size_t)len + 1);
    size_t read = fread(buf, 1, (size_t)len, fp);
    fclose(fp);
    buf[read] = '\0';
    return buf;
}

char *skill_strip_yaml_frontmatter(const char *path) {
    char *contents = read_file_to_string(path);
    if (!contents) return NULL;

    // Frontmatter must start with "---" on the first line (allow leading
    // whitespace? The skill parser doesn't, so we don't either).
    if (strncmp(contents, "---", 3) != 0 ||
        (contents[3] != '\n' && contents[3] != '\r')) {
        return contents;
    }

    // Find the closing "---" on its own line.
    char *p = contents + 3;
    while (*p == '\r' || *p == '\n') p++;

    while (*p) {
        char *line_start = p;
        char *line_end = strchr(p, '\n');
        size_t line_len = line_end ? (size_t)(line_end - line_start) : strlen(line_start);

        // Trim trailing \r for CRLF files.
        size_t check_len = line_len;
        if (check_len > 0 && line_start[check_len - 1] == '\r') check_len--;

        if (check_len == 3 && strncmp(line_start, "---", 3) == 0) {
            // Skip past the closing delimiter and any trailing newlines.
            char *body = line_end ? line_end + 1 : line_start + line_len;
            char *body_dup = str_dup(body);
            spm_free(contents);
            return body_dup;
        }

        if (!line_end) break;
        p = line_end + 1;
    }

    // Unterminated frontmatter — fall back to returning the original contents.
    log_debug("Unterminated frontmatter in %s", path);
    return contents;
}

// Check if a directory contains SKILL.md
static Skill *check_skill_dir(const char *dir_path) {
    char *skill_file = path_join(dir_path, "SKILL.md");
    Skill *skill = NULL;

    if (file_exists(skill_file)) {
        skill = parse_skill_file(skill_file);
    }

    spm_free(skill_file);
    return skill;
}

// Recursively search for SKILL.md files
static void find_skills_recursive(const char *base_dir, SkillList *list, int depth) {
    if (depth > 5) return;  // Limit recursion depth

    DIR *dir = opendir(base_dir);
    if (!dir) return;

    struct dirent *entry;

    while ((entry = readdir(dir)) != NULL) {
        if (entry->d_name[0] == '.') continue;  // Skip hidden files

        char *full_path = path_join(base_dir, entry->d_name);
        struct stat st;

        if (stat(full_path, &st) == 0 && S_ISDIR(st.st_mode)) {
            // Check if this directory contains SKILL.md
            Skill *skill = check_skill_dir(full_path);
            if (skill) {
                skill_list_add(list, skill);
            } else {
                // Recurse into subdirectory
                find_skills_recursive(full_path, list, depth + 1);
            }
        }

        spm_free(full_path);
    }

    closedir(dir);
}

SkillList *discover_skills(const char *base_dir) {
    SkillList *list = spm_malloc(sizeof(SkillList));
    list->skills = NULL;
    list->count = 0;
    list->capacity = 0;

    // Check root level first
    Skill *root_skill = check_skill_dir(base_dir);
    if (root_skill) {
        skill_list_add(list, root_skill);
    }

    // Check known skill paths
    for (int i = 0; SKILL_SEARCH_PATHS[i] != NULL; i++) {
        char *search_path = path_join(base_dir, SKILL_SEARCH_PATHS[i]);

        if (dir_exists(search_path)) {
            log_debug("Searching for skills in: %s", search_path);
            find_skills_recursive(search_path, list, 0);
        }

        spm_free(search_path);
    }

    // If nothing found in known paths, search from root
    if (list->count == 0) {
        log_debug("No skills in known paths, searching recursively from root");
        find_skills_recursive(base_dir, list, 0);
    }

    return list;
}

void skill_print(const Skill *skill) {
    if (!skill) return;

    // Use color if outputting to a terminal
    int use_color = isatty(fileno(stdout));

    if (use_color) {
        printf("  \033[1;34m%s\033[0m", skill->name);  // Bold blue
    } else {
        printf("  %s", skill->name);
    }

    if (skill->description) {
        printf(" - %s", skill->description);
    }
    printf("\n");
}

void skill_list_print(const SkillList *list) {
    if (!list || list->count == 0) {
        printf("  (no skills found)\n");
        return;
    }

    for (int i = 0; i < list->count; i++) {
        skill_print(&list->skills[i]);
    }
}

File: src/skill.rs
// Skill discovery and SKILL.md frontmatter parsing.
//
// SKILL.md format:
//   ---
//   name: skill-name
//   description: Some description
//   ---
//   <body>
//
// Only `name` and `description` are parsed — everything else is ignored.

use crate::os;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct Skill {
    pub name: String,
    pub description: Option<String>,
    /// Directory containing the SKILL.md.
    pub path: PathBuf,
    /// Path to the SKILL.md file itself.
    pub skill_file: PathBuf,
}

/// Search paths checked in order when looking inside a *package* for the
/// skills it ships. Intentionally narrow: `skills/` is the convention, and
/// the recursive fallback below catches anything off-convention. We do not
/// scan agent-specific directories like `.agents/skills`, `.claude/skills`,
/// `.cursor/skills`, etc. — those are consumer-side install destinations,
/// so a project that committed its installed third-party skills would have
/// them mistakenly republished.
const SKILL_SEARCH_PATHS: &[&str] = &["skills"];

/// Parse SKILL.md frontmatter. Returns None on read error or when no name
/// can be derived from frontmatter or the parent directory name.
pub fn parse_skill_file(path: &Path) -> Option<Skill> {
    let contents = match os::read_to_string(path) {
        Ok(s) => s,
        Err(_) => {
            crate::log::debug(&format!("Cannot open: {}", path.display()));
            return None;
        }
    };

    let mut name: Option<String> = None;
    let mut description: Option<String> = None;
    let mut in_frontmatter = false;
    let mut closed_frontmatter = false;

    for line in contents.lines() {
        let trimmed = line.trim();
        if trimmed == "---" {
            if !in_frontmatter {
                in_frontmatter = true;
                continue;
            }
            closed_frontmatter = true;
            break;
        }
        if !in_frontmatter {
            continue;
        }
        if let Some(idx) = trimmed.find(':') {
            let key = trimmed[..idx].trim();
            let mut value = trimmed[idx + 1..].trim();
            // Strip matching quotes.
            if value.len() >= 2 {
                let first = value.as_bytes()[0];
                let last = value.as_bytes()[value.len() - 1];
                if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
                    value = &value[1..value.len() - 1];
                }
            }
            match key {
                "name" => name = Some(value.to_string()),
                "description" => description = Some(value.to_string()),
                _ => {}
            }
        }
    }

    // If no name from frontmatter, fall back to the parent directory name.
    if name.is_none() {
        if let Some(parent) = path.parent() {
            if let Some(dir_name) = parent.file_name() {
                if let Some(s) = dir_name.to_str() {
                    name = Some(s.to_string());
                }
            }
        }
    }

    let name = name?;
    let dir = path
        .parent()
        .map(|p| p.to_path_buf())
        .unwrap_or_else(|| PathBuf::from("."));

    // We don't strictly need closed_frontmatter; the C code is also lenient.
    let _ = closed_frontmatter;

    Some(Skill {
        name,
        description,
        path: dir,
        skill_file: path.to_path_buf(),
    })
}

/// Return the body of a markdown file with leading YAML frontmatter stripped.
/// Returns the full file contents when no frontmatter is present, or None on
/// read error. Mirrors `skill_strip_yaml_frontmatter` from skill.c.
pub fn strip_yaml_frontmatter(path: &Path) -> Option<String> {
    let contents = os::read_to_string(path).ok()?;
    let bytes = contents.as_bytes();
    // Must begin with "---" followed by \n or \r.
    if bytes.len() < 4 || &bytes[..3] != b"---" || (bytes[3] != b'\n' && bytes[3] != b'\r') {
        return Some(contents);
    }
    // Walk lines looking for a sole "---" terminator.
    let mut i = 3;
    while i < bytes.len() && (bytes[i] == b'\r' || bytes[i] == b'\n') {
        i += 1;
    }
    while i < bytes.len() {
        let line_start = i;
        let line_end = bytes[line_start..]
            .iter()
            .position(|&b| b == b'\n')
            .map(|j| line_start + j);
        let raw_end = line_end.unwrap_or(bytes.len());
        let mut check_end = raw_end;
        if check_end > line_start && bytes[check_end - 1] == b'\r' {
            check_end -= 1;
        }
        if check_end - line_start == 3 && &bytes[line_start..line_start + 3] == b"---" {
            // Skip past the closing delimiter line.
            let body_start = match line_end {
                Some(le) => le + 1,
                None => raw_end,
            };
            return Some(contents[body_start..].to_string());
        }
        match line_end {
            Some(le) => i = le + 1,
            None => break,
        }
    }
    crate::log::debug(&format!("Unterminated frontmatter in {}", path.display()));
    Some(contents)
}

/// Walk a directory tree (max depth 5) for SKILL.md files and parse each.
fn find_skills_recursive(base: &Path, out: &mut Vec<Skill>, depth: u32) {
    if depth > 5 {
        return;
    }
    let entries = match os::read_dir(base) {
        Ok(v) => v,
        Err(_) => return,
    };
    for name in entries {
        if name.starts_with('.') {
            continue;
        }
        let child = base.join(&name);
        let kind = match os::metadata(&child) {
            Ok(m) => m.kind,
            Err(_) => continue,
        };
        if kind != os::FileKind::Dir {
            continue;
        }
        let skill_md = child.join("SKILL.md");
        if os::is_file(&skill_md) {
            if let Some(skill) = parse_skill_file(&skill_md) {
                out.push(skill);
                continue;
            }
        }
        find_skills_recursive(&child, out, depth + 1);
    }
}

/// Find all skills in a directory tree. Checks the root first, then each
/// path in `SKILL_SEARCH_PATHS`, then walks the whole tree if nothing was
/// found in the known locations.
pub fn discover_skills(base_dir: &Path) -> Vec<Skill> {
    let mut out = Vec::new();

    // Root-level SKILL.md.
    let root_md = base_dir.join("SKILL.md");
    if os::is_file(&root_md) {
        if let Some(s) = parse_skill_file(&root_md) {
            out.push(s);
        }
    }

    for sub in SKILL_SEARCH_PATHS {
        let search = base_dir.join(sub);
        if os::is_dir(&search) {
            crate::log::debug(&format!("Searching for skills in: {}", search.display()));
            find_skills_recursive(&search, &mut out, 0);
        }
    }

    if out.is_empty() {
        crate::log::debug("No skills in known paths, searching recursively from root");
        find_skills_recursive(base_dir, &mut out, 0);
    }

    out
}

pub fn print(skill: &Skill) {
    match &skill.description {
        Some(d) => crate::log::info(&format!("  {} - {}", skill.name, d)),
        None => crate::log::info(&format!("  {}", skill.name)),
    }
}

pub fn print_list(list: &[Skill]) {
    if list.is_empty() {
        crate::log::info("  (no skills found)");
        return;
    }
    for s in list {
        print(s);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tmpdir() -> PathBuf {
        let p = std::env::temp_dir().join(format!(
            "rosie-skill-test-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        std::fs::create_dir_all(&p).unwrap();
        p
    }

    #[test]
    fn parse_frontmatter() {
        let dir = tmpdir();
        let md = dir.join("foo").join("SKILL.md");
        std::fs::create_dir_all(md.parent().unwrap()).unwrap();
        std::fs::write(
            &md,
            "---\nname: my-skill\ndescription: a test skill\n---\n\n# my-skill\n",
        )
        .unwrap();
        let s = parse_skill_file(&md).expect("should parse");
        assert_eq!(s.name, "my-skill");
        assert_eq!(s.description.as_deref(), Some("a test skill"));
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn strip_frontmatter() {
        let dir = tmpdir();
        let md = dir.join("S.md");
        std::fs::write(&md, "---\nname: x\n---\n\n# body\n").unwrap();
        let body = strip_yaml_frontmatter(&md).unwrap();
        assert_eq!(body, "\n# body\n");
        std::fs::remove_dir_all(&dir).ok();
    }
}
